File: src/CelSerEngine.WpfBlazor/Components/AgGrid/VirtualizedAgGrid.razor.js
let gridApi;
let dotNetHelper;
let verticalViewport;
const itemHeight = 25;
const maxDivHeight = 32000000;
let totalItemCount = 0;

async function initVirtualizedAgGrid(_dotNetHelper, _gridOptions) {
    dotNetHelper = _dotNetHelper;

    const loadingOverlay = class CustomLoadingOverlay {
        init(params) {
            this.eGui = document.createElement('div');

            if (params.isScanning) {
                this.eGui.innerHTML = `<div class="spinner-border m-auto" role="status"></div>`;
            }
            else {
                this.eGui.innerHTML = ``;
            }
        }

        getGui() {
            return this.eGui;
        }

        refresh(params) {
            return false;
        }
    };

    const gridOptions = {
        defaultColDef: {
            suppressMovable: true,
            sortable: false,
            flex: 1
        },
        animateRows: false,
        loadingOverlayComponent: loadingOverlay,
        loadingOverlayComponentParams: {
            isScanning: false,
        },
        getRowId: (params) => params.data.RowId,
        onModelUpdated: (params) => {
            const nodesToSelect = [];
            params.api.forEachNode((node) => {
                if (node.data && node.data.IsSelected) {
                    nodesToSelect.push(node);
                }
            });
            params.api.setNodesSelected({ nodes: nodesToSelect, newValue: true });
        },
        onRowDoubleClicked: onRowDoubleClicked
    };
    
    if (_gridOptions.getRowStyleFunc != null) {
        gridOptions["getRowStyle"] = params => {
            if (params.node.data) {
                const getRowStyleFunc = new Function("itemParam", `return ${_gridOptions.getRowStyleFunc}(itemParam);`);

                return getRowStyleFunc(params.node.data.Item);
            }
        };
    }

    gridOptions["columnDefs"] = generateColumnDefs(_gridOptions.columnDefs);
    gridOptions["rowSelection"] = _gridOptions.rowSelection.toLowerCase();

    const myGridElement = document.querySelector('#scanResultsGrid');
    gridApi = agGrid.createGrid(myGridElement, gridOptions);

    myGridElement.dataset.previousHeight = myGridElement.clientHeight;
    new ResizeObserver(onResize).observe(myGridElement);
    const fakeVscroll = document.querySelector(".ag-body-vertical-scroll");
    const clone = fakeVscroll.cloneNode(true);
    fakeVscroll.replaceWith(clone);
    clone.classList.remove("ag-hidden");
    verticalViewport = document.querySelector(".ag-body-vertical-scroll-viewport");
    handleEvents();
    await dotNetHelper.invokeMethodAsync("InitItemCount", getVisibleRowCount());
}

function handleEvents() {
    verticalViewport.addEventListener("scroll", async () => await onScroll());
    document.querySelector(".ag-body-vertical-scroll-container").style.height = `${Math.min(totalItemCount * itemHeight, maxDivHeight)}px`;
    document.querySelector(".ag-body-viewport").addEventListener("wheel", function (e) {
        if (Math.abs(e.deltaY) > 0) {
            const deltaSign = Math.sign(e.deltaY);
            verticalViewport.scrollBy(0, itemHeight * deltaSign);
            this.scrollTop = (this.scrollHeight - this.clientHeight) * deltaSign;
            e.preventDefault();
        }
    });

    document.querySelector(".ag-center-cols-container").addEventListener("click", async function (e) {
        const clickedRow = e.target.parentElement;
        if (clickedRow.classList.contains("ag-row")) {
            const rowId = clickedRow.getAttribute("row-id");

            if (!e.shiftKey && !e.ctrlKey) {
                await dotNetHelper.invokeMethodAsync("ClearSelectedItems", rowId);
            }

            if (e.shiftKey) {
                e.preventDefault();
                await dotNetHelper.invokeMethodAsync("SelectTillItemAsync", rowId);
                gridApi.forEachNode(async (node) => node.selectThisNode(await dotNetHelper.invokeMethodAsync("IsItemSelectedAsync", node.data.RowId)));
                return;
            }

            const rowIsSelected = gridApi.getRowNode(rowId).isSelected();
            if (rowIsSelected) {
                dotNetHelper.invokeMethodAsync("AddSelectedItemAsync", rowId);
            }
            else {
                dotNetHelper.invokeMethodAsync("RemoveSelectedItemAsync", rowId);
            }
        }
    });
}

async function onResize(entry) {
    if (totalItemCount == 0)
        return;

    const currentHeight = entry[0].target.clientHeight;
    const previousHeight = entry[0].target.dataset.previousHeight || 0;

    if (currentHeight > previousHeight) {
        if (currentHeight - previousHeight >= itemHeight) {
            entry[0].target.dataset.previousHeight = currentHeight;
            const visibleRowCount = getVisibleRowCount();
            await loadItemsIntoGrid(getStartIndex(visibleRowCount), visibleRowCount, true);
        }
    } else if (currentHeight < previousHeight) {
        if (previousHeight - currentHeight >= itemHeight) {
            entry[0].target.dataset.previousHeight = currentHeight;
        }
    }
}

async function onRowDoubleClicked(row) {
    await dotNetHelper.invokeMethodAsync('OnRowDoubleClickedDispatcherAsync', row.data.RowId);
}

function showLoadingOverlay(showSpinner = true) {
    gridApi.setGridOption("loadingOverlayComponentParams", { isScanning: showSpinner });
    gridApi.showLoadingOverlay();
}

function resetGrid() {
    gridApi.setGridOption("rowData", []);
    showLoadingOverlay(false);
}

async function itemsChanged(totalCount) {
    totalItemCount = totalCount;
    document.querySelector(".ag-body-vertical-scroll-container").style.height = `${Math.min(totalItemCount * itemHeight, maxDivHeight)}px`;
    const visibleRowCount = getVisibleRowCount();
    const startIndex = getStartIndex(visibleRowCount);
    await loadItemsIntoGrid(startIndex, visibleRowCount, false);
}

async function loadItemsIntoGrid(startIndex, rowCount, forceFetchItems) {
    const visibleItems = await dotNetHelper.invokeMethodAsync('GetItemsAsync', startIndex, rowCount, forceFetchItems)
        .then(x => JSON.parse(x));
    gridApi.setGridOption("rowData", visibleItems);

    return visibleItems;
}

function getStartIndex(visibleRowCount) {
    let startIndex = -1;

    if (verticalViewport.clientHeight + verticalViewport.scrollTop == maxDivHeight) {
        startIndex = totalItemCount - visibleRowCount;
    }
    else if ((totalItemCount * itemHeight) >= maxDivHeight) {
        startIndex = Math.floor(((totalItemCount * itemHeight) / maxDivHeight) * verticalViewport.scrollTop / itemHeight);
    }
    else {
        startIndex = Math.floor(verticalViewport.scrollTop / itemHeight);
    }

    return startIndex;
}

function getVisibleRowCount() {
    return Math.ceil(verticalViewport.clientHeight / itemHeight);
}

var lastStartIndex = -1;
async function onScroll() {
    const visibleRowCount = getVisibleRowCount();
    let startIndex = getStartIndex(visibleRowCount);

    if (lastStartIndex == startIndex) {
        return;
    }

    lastStartIndex = startIndex;
    const visibleItems = await loadItemsIntoGrid(startIndex, visibleRowCount, true);
}

function generateColumnDefs(columnDefs) {
    const newColDefs = [];
    for (const columnDef of columnDefs) {
        if (columnDef.isArray) {
            // create cols from array items
            for (var i = 0; i < columnDef.arraySize; i++) {
                const captureIndex = i; // this could be done using the p parameter in the valueGetter, but this is way easier.
                newColDefs.push({
                    valueGetter: p => p.data.Item[columnDef.field][captureIndex],
                    headerName: `${columnDef.headerName} ${i}`,
                });
            }
        }
        else {
            newColDefs.push({
                field: `Item.${columnDef.field}`,
                headerName: columnDef.headerName ?? columnDef.field,
            });
        }
    }

    return newColDefs;
}

function updateColumnDefs(columnDefs) {
    gridApi.setGridOption("columnDefs", generateColumnDefs(columnDefs)); 
}

export { initVirtualizedAgGrid, itemsChanged, showLoadingOverlay, resetGrid, updateColumnDefs }

File: src/CelSerEngine.WpfBlazor/Components/AgGrid/VirtualizedAgGrid.razor.cs
using Microsoft.AspNetCore.Components;
using Microsoft.JSInterop;
using System.Globalization;
using System.Text.Json;

namespace CelSerEngine.WpfBlazor.Components.AgGrid;

public partial class VirtualizedAgGrid<TSource, TDisplay> : ComponentBase, IAsyncDisposable
{
    public HashSet<string> SelectedItems { get; set; } = [];

    [Inject]
    private IJSRuntime JSRuntime { get; set; } = default!;

    [Inject]
    private ThemeManager ThemeManager { get; set; } = default!;

    /// <summary>
    /// Gets or sets the fixed item source.
    /// </summary>
    [Parameter]
    public ICollection<TSource> Items { get; set; } = default!;

    [Parameter]
    public Func<int, int, Task<(ICollection<TSource> items, int totalCount)>>? ServerItems { get; set; }

    [Parameter]
    public Func<TSource, string> GetRowId { get; set; } = default!;

    [Parameter]
    public GridOptions GridOptions { get; set; } = default!;

    [Parameter]
    public EventCallback<TSource> OnRowDoubleClicked { get; set; }

    private int TotalItemsCount { get; set; }

    private CultureInfo _cultureInfo = new("en-US");
    private IJSObjectReference? _module;
    private DotNetObjectReference<VirtualizedAgGrid<TSource, TDisplay>>? _dotNetHelper;
    private int _lastStartIndex = 0;
    private int _lastItemCount = 0;
    private (ICollection<TSource> items, int totalItemCount)? _lastServerItems;
    private bool _disposed;
    private bool _isUpdating;

    /// <inheritdoc />
    protected override async Task OnAfterRenderAsync(bool firstRender)
    {
        if (firstRender)
        {
            _dotNetHelper = DotNetObjectReference.Create(this);
            _module = await JSRuntime.InvokeAsync<IJSObjectReference>("import", "./Components/AgGrid/VirtualizedAgGrid.razor.js");
            await _module.InvokeVoidAsync("initVirtualizedAgGrid", _dotNetHelper, GridOptions);
        }
    }

    public async Task UpdateColumnDefs()
    {
        await _module!.InvokeVoidAsync("updateColumnDefs", (object)GridOptions.ColumnDefs);
    }

    public async Task ApplyDataAsync()
    {
        if (_disposed || _isUpdating)
            return;

        _isUpdating = true;
        TotalItemsCount = Items.Count;

        if (ServerItems != null)
        {
            _lastServerItems ??= await ServerItems(_lastStartIndex, _lastItemCount);
            TotalItemsCount = _lastServerItems.Value.totalItemCount;
        }

        await _module!.InvokeVoidAsync("itemsChanged", TotalItemsCount);
        _isUpdating = false;
    }

    public async Task ShowScanningOverlayAsync()
    {
        await _module!.InvokeVoidAsync("showLoadingOverlay");
    }

    public async Task ResetGridAsync()
    {
        await _module!.InvokeVoidAsync("resetGrid");
    }

    public IEnumerable<TSource> GetVisibleItems()
    {
        if (ServerItems != null && _lastServerItems != null)
        {
            return _lastServerItems.Value.items;
        }

        return Items.Skip(_lastStartIndex).Take(_lastItemCount);
    }

    [JSInvokable]
    public Task OnRowDoubleClickedDispatcherAsync(string rowId)
    {
        if (OnRowDoubleClicked.HasDelegate)
        {
            var doubleClickedRow = Items.Single(x => GetRowId(x) == rowId);
            OnRowDoubleClicked.InvokeAsync(doubleClickedRow);
        }

        return Task.CompletedTask;
    }

    [JSInvokable]
    public async Task<string> GetItemsAsync(int startIndex, int amount, bool forceFetchItems)
    {
        _lastStartIndex = startIndex;
        _lastItemCount = amount;

        IEnumerable<TSource> visibleItems;

        if (ServerItems != null)
        {
            if ((forceFetchItems || _lastServerItems == null))
            {
                _lastServerItems = await ServerItems(startIndex, amount);
            }

            visibleItems = _lastServerItems.Value.items;
        }
        else
        {
            visibleItems = Items.Skip(startIndex).Take(amount);
        }

        return JsonSerializer.Serialize(visibleItems.Select(x => new
        {
            Item = (TDisplay)Activator.CreateInstance(typeof(TDisplay), x)!,
            IsSelected = SelectedItems.Contains(GetRowId(x)),
            RowId = GetRowId(x)
        }));
    }

    [JSInvokable]
    public Task SelectTillItemAsync(string selectTillItem)
    {
        var isSelecting = false;
        var lastAddress = selectTillItem;
        var firstAddress = SelectedItems.LastOrDefault();

        if (firstAddress == null)
        {
            SelectedItems.Add(lastAddress);

            return Task.CompletedTask;
        }

        if (firstAddress == lastAddress)
            return Task.CompletedTask;

        if (GridOptions.RowSelection == RowSelection.Single)
        {
            SelectedItems.Clear();
            SelectedItems.Add(lastAddress);

            return Task.CompletedTask;
        }

        foreach (var item in Items)
        {
            var address = GetRowId(item);
            if (address == firstAddress || address == lastAddress)
            {
                isSelecting = !isSelecting;

                if (!isSelecting)
                {
                    SelectedItems.Add(address);
                    break;
                }
            }

            if (isSelecting)
                SelectedItems.Add(address);
        }

        return Task.CompletedTask;
    }

    [JSInvokable]
    public Task<bool> IsItemSelectedAsync(string item)
    {
        return Task.FromResult(SelectedItems.Contains(item));
    }

    [JSInvokable]
    public Task ClearSelectedItems()
    {
        SelectedItems.Clear();

        return Task.CompletedTask;
    }

    [JSInvokable]
    public Task AddSelectedItemAsync(string item)
    {
        if (GridOptions.RowSelection == RowSelection.Single)
        {
            SelectedItems.Clear();
        }

        SelectedItems.Add(item);

        return Task.CompletedTask;
    }

    [JSInvokable]
    public Task RemoveSelectedItemAsync(string item)
    {
        SelectedItems.Remove(item);

        return Task.CompletedTask;
    }

    [JSInvokable]
    public void InitItemCount(int itemCount)
    {
        _lastItemCount = itemCount;
    }

    /// <inheritdoc />
    public async ValueTask DisposeAsync()
    {
        _disposed = true;
        _dotNetHelper?.Dispose();

        if (_module != null)
            await _module.DisposeAsync();

        GC.SuppressFinalize(this);
    }
}
